### PR TITLE
LP order sizing

### LINKED ISSUE #6955: Refactor liquidity order lp orders sizing
Implement the changes introduced by the following specs changes: https://github.com/vegaprotocol/specs/pull/1438

- [ ] Remove scaling by probability when implying LP volumes.
- [ ] Only change the price if it’s outside the new “valid LP price range” - move it to the bound in that case.

### PR DESCRIPTION
**Pull request by witgaw**
Implement part of the spec changes introduced by https://github.com/vegaprotocol/specs/pull/1438:
- Extend API with the new market field: LP price range.
- Use the new field to move automated LP orders falling outside of the range so that they end up on the bound (previously price monitoring bounds were used for that)/
- LP orders no longer scaled by probability of trading.

Closes #6955 

### https://github.com/vegaprotocol/vega/pull/7074
core/integration/features/fees/trading-fees-pdp.feature
core/integration/features/fees/setting-fee-and-rewarding-lps.feature
core/integration/features/fees/amend-fees.feature
core/integration/features/rewards/multi_asset_multi_market_fees_rewards.feature
core/integration/features/rewards/infra_fees.feature
core/integration/features/rewards/asset_fees_rewards.feature
core/integration/features/accounts/2943-distressed-trader-has-general-balance.feature
core/integration/features/auctions/5460-market-order-cannot-trade-due-to-auction.feature

### https://github.com/vegaprotocol/vega/pull/7057

core/integration/features/verified/0038-OLIQ-liquidity_provision_creation.feature
core/integration/features/verified/0038-OLIQ-verified-complex-shape-order-sizes.feature
core/integration/features/verified/0042-LIQF-fees_rewards.feature
core/integration/features/verified/0042-LIQF-fees_rewards_with_decimal.feature
core/integration/features/verified/0042-LIQF-fees_rewards_large_distribution_time.feature
core/integration/features/verified/0042-LIQF-fees_rewards_growing_market.feature
core/integration/features/verified/0042-LIQF-fees_rewards_VirtualStake.feature
core/integration/features/verified/0042-LIQF-fees_rewards_multi_lps.feature

### https://github.com/vegaprotocol/vega/pull/7109

core/integration/features/liquidity-provision/liquidity_provision_simple_pdp.feature
core/integration/features/liquidity-provision/mid-price-issue-amend.feature
core/integration/features/liquidity-provision/mid-price-issue.feature
core/integration/features/liquidity-provision/automatic-adjustment-to-order-offsets-and-volumes.feature
core/integration/features/liquidity-provision/verify-simple-shape-order-sizes.feature
core/integration/features/liquidity-provision/verify-simple-shape-order-sizes-best-bid-and-ask.feature
core/integration/features/liquidity-provision/proposal-no-lp.feature
core/integration/features/liquidity-provision/unexpected-auction.feature
core/integration/features/liquidity-provision/lp-distressed-closeout.feature
core/integration/features/liquidity-provision/lp-distressed-closeout-pdp.feature
core/integration/features/liquidity-provision/verify-CcySiska.feature
core/integration/features/liquidity-provision/lp-margin-post-auction.feature
core/integration/features/liquidity-provision/margin-release-cancel.feature
core/integration/features/liquidity-provision/bug-5110.feature
core/integration/features/liquidity-provision/bug-5147.feature

### https://github.com/vegaprotocol/vega/pull/7086

core/integration/features/margin/596-Opened-positions-traded-out-but-there-is-still-money-sitting-in-a-margin-account.feature
core/integration/features/margin/596-margin-account-non-zero-after-trading-out-market-order.feature
core/integration/features/margin/598-open-position-with-zero-margin.feature
core/integration/features/margin/614-margin-calculations-fixes.feature
core/integration/features/position_tracking/verified-positions-resolution-5-lognormal.feature
core/integration/features/position_tracking/verified-positions-resolution-5-lognormal_pdp.feature
core/integration/features/settlement/mark_to_market_settlement_neg_pdp.feature
core/integration/features/settlement/mark_to_market_interval_positions.feature
core/integration/features/settlement/mark_to_market_settlement_pdp.feature
core/integration/features/settlement/markt_to_market_interval.feature
core/integration/features/price_monitoring/volume-near-price-mon-bounds.feature
core/integration/features/price_monitoring/3362-failing-system-tests.feature
core/integration/features/price_monitoring/volume-near-price-mon-bounds-2.feature

### https://github.com/vegaprotocol/vega/pull/7115

core/integration/features/verified/closeout-cascade.feature
core/integration/features/verified/closeout-lognormal.feature
core/integration/features/verified/Funding-insurance_pool.feature
core/integration/features/verified/Issue-6004-ProbOfTrading_Refactoring.feature
core/integration/features/verified/liquidity-provision-bond-account-dp.feature
core/integration/features/verified/liquidity-provision-bond-account.feature
core/integration/features/verified/Margin-Levels.feature
core/integration/features/verified/risk-parameters-ranges-test.feature
core/integration/features/verified/risk-parameters-sigma-test.feature
core/integration/features/verified/Slippage-Calc-in-Margin-Level.feature
core/integration/features/verified/Test-PnL.feature
core/integration/features/verified/test-tau-scaling.feature

### https://github.com/vegaprotocol/vega/pull/7123

core/integration/features/verified/0011-MARA-006.feature
core/integration/features/verified/0011-MARA-016017.feature
core/integration/features/verified/0011-MARA-change_in_margin figures.feature
core/integration/features/verified/0019-MCAL-margin_calculator-risk_parameters.feature
core/integration/features/verified/0029-FEES-fees.feature
core/integration/features/verified/0029-FEES-trading_fees.feature
core/integration/features/verified/0034-PROB-liquidity_measure.feature
core/integration/features/verified/0070-MKTD-market-decimal-places.feature

### https://github.com/vegaprotocol/vega/pull/7129

core/integration/features/verified/0006-POSI-AC-008.feature
core/integration/features/verified/0015-INSR-insurance_pool_balance_test.feature
core/integration/features/verified/0026-AUCT-auction_interaction.feature
core/integration/features/verified/0032-PRIM-price_monitoring_horizon.feature

### https://github.com/vegaprotocol/vega/pull/7135

core/integration/features/verified/0035-LIQM-liquidity_monitoring.feature

### CODE REVIEW

File: core/execution/liquidity_provision.go [hunk 1]
```diff
@@ -494,205 +494,101 @@ func (m *Market) rollBackMargin(
...
+		return num.UintZero(), ErrUnableToReprice
 	}
+	minLpPrice, maxLpPrice, err := m.getValidLPVolumeRange()
 	if err != nil {
-		return num.UintZero(), nil, ErrUnableToReprice
+		return num.UintZero(), err
 	}
-	return m.adjustPriceRange(po, side, price)
+	adjustedPrice := m.adjustPrice(side, refPrice, offset, minLpPrice, maxLpPrice)
+	return adjustedPrice, nil
```
Comment (jeremyletang) [inline core/execution/liquidity_provision.go:522]:
```suggestion
	return m.adjustPrice(side, refPrice, offset, minLpPrice, maxLpPrice), nil
```

File: core/execution/liquidity_provision.go [hunk 2]
```diff
@@ -494,205 +494,101 @@ func (m *Market) rollBackMargin(
...
+	ub, _ := num.UintFromDecimal(ubD)
+
+	// floor at 1 to avoid non-positive value
+	if lb.IsNegative() || lb.IsZero() {
+		lb = m.minValidPrice()
+	}
+
+	if lb.GTE(ub) {
+		// if we ended up with overlapping upper and lower bound we set the upper bound to lower bound plus one.
+		ub = num.UintZero().Add(lb, num.UintOne())
```
Comment (jeremyletang) [inline core/execution/liquidity_provision.go:588]:
```suggestion
		ub = ub.Add(lb, num.UintOne())
```

File: core/liquidity/supplied/engine.go [hunk 3]
```diff
@@ -128,109 +127,72 @@ func (e *Engine) CalculateSuppliedLiquidity(
...
-		sum = sum.Add(proportion)
-		validatedProportions = append(validatedProportions, proportion)
-		probs = append(probs, prob)
+		prop := num.DecimalFromUint(num.NewUint(uint64(o.Details.Proportion)))
+		proportionsD = append(proportionsD, prop)
+		sum = sum.Add(prop)
 	}
 	if sum.IsZero() {
+		// TODO: This can probably be removed now and a lot of upstream code can be simplified (no error handling)
 		return ErrNoValidOrders
```
Comment (jeremyletang) [inline core/liquidity/supplied/engine.go:190]: let's do it then?

**Comment by vega-paul** [comment]
```{"system-tests":"vega_7014"}```

File: core/liquidity/supplied/engine.go (hunk 3 above)
Comment (witgaw) [reply core/liquidity/supplied/engine.go]: Done now

**Review by davidsiska-vega — DISMISSED** [review]

File: commands/proposal_submission.go [hunk 4]
```diff
@@ -369,6 +369,18 @@ func checkNewMarketChanges(change *protoTypes.ProposalTerms_NewMarket) Errors {
 		errs.AddForProperty("proposal_submission.terms.change.new_market.changes.position_decimal_places", ErrMustBeWithinRange7)
 	}
 
+	lppr, err := num.DecimalFromString(changes.LpPriceRange)
+	if err != nil {
+		errs.AddForProperty("proposal_submission.terms.change.new_market.changes.lp_price_range", ErrIsNotValidNumber)
+	} else {
+		if lppr.IsNegative() || lppr.IsZero() {
```
Comment (jeremyletang) [inline commands/proposal_submission.go:376]: could be
```suggestion
		if lppr.Cmp(num.DecimalZero()) <= 0 {
```

not specifically better I suppose though

File: commands/proposal_submission.go [hunk 5]
```diff
@@ -395,6 +407,17 @@ func checkUpdateMarketChanges(change *protoTypes.ProposalTerms_UpdateMarket) Err
 	}
 
 	changes := change.UpdateMarket.Changes
+	lppr, err := num.DecimalFromString(changes.LpPriceRange)
+	if err != nil {
+		errs.AddForProperty("proposal_submission.terms.change.update_market.changes.lp_price_range", ErrIsNotValidNumber)
+	} else {
+		if lppr.IsNegative() || lppr.IsZero() {
```
Comment (jeremyletang) [inline commands/proposal_submission.go:414]: same here

File: core/execution/liquidity_provision.go [hunk 6]
```diff
@@ -173,25 +173,9 @@ func (m *Market) SubmitLiquidityProvision(
...
-			logging.String("market-id", m.GetID()),
-			logging.String("party", party),
-			logging.Error(err),
-		)
-		// at this point, we were able to take the bond from the party
-		// but were not able to generate the orders
-		// this is likely due to the market not being ready and the liquidity
-		// engine not being able to price the orders
-		// we do not want to rollback anything then
-		needsBondRollback = false
```
[the remark this answers is not in the document]
Comment (witgaw) [reply core/execution/liquidity_provision.go:188]: It's still the same though, all that's changed is the error message.

Used to be: `m.getBestStaticPricesDecimal()` => if err => needsBondRollback=false, needsCancel=false => return
Now:  `m.getValidLPVolumeRange()` (which depends on best bid and ask so same behaviour) => if err => needsBondRollback=false, needsCancel=false => return

It's just the diff that's making it look weird

File: commands/proposal_submission.go (hunk 4 above)
Comment (witgaw) [reply commands/proposal_submission.go:376]: I think mine is more readable for ppl not that familiar with the library

**Review by davidsiska-vega — DISMISSED** [review]

File: commands/proposal_submission.go (hunk 5 above)
Comment (EVODelavega) [reply commands/proposal_submission.go:414]: Not a fan of `else { if {`, perhaps using `else if`?

Comment (witgaw) [reply commands/proposal_submission.go]: Sure, why not, changed now.

File: core/execution/special_orders.go [hunk 7]
```diff
@@ -159,35 +159,24 @@ func (m *Market) repriceAllSpecialOrders(
...
-
 	// now we have all the re-submitted pegged orders and the
 	// parked pegged orders from before
 	// we can call liquidityUpdate, which is going to give us the
 	// actual updates to be done on liquidity orders
-	bestBidPrice, bestAskPrice, err := m.getBestStaticPricesDecimal()
+	minLpPrice, maxLpPrice, err := m.getValidLPVolumeRange()
 	if err != nil {
-		m.log.Debug("could not get one of the static mid prices",
+		m.log.Debug("could not get valid LP volume range",
```
Comment (EVODelavega) [inline core/execution/special_orders.go:173]: I get that this is what we had before, but it looks kind of off to me. If we get an error trying to get valid LP min/max prices, should we even continue?

**Review by EVODelavega — APPROVED** [review]

**Review by davidsiska-vega — APPROVED** [review]

**Review by mattrussell36 — APPROVED** [review]

**Pull request merged**